### PR TITLE
osd: remove the lock heartbeat_update_lock, and change heatbeat_need_…

### PR DESCRIPTION
**Pull request by liupan1111**
…update to atomic_bool

Signed-off-by: Pan Liu <pan.liu@istuary.com>

### CODE REVIEW

**Comment by liupan1111** [comment]
@tchaikov or @liewegas , please help take a look.

File: src/osd/OSD.h [hunk 1]
```diff
@@ -1631,8 +1631,7 @@ class OSD : public Dispatcher,
   map<int, int> debug_heartbeat_drops_remaining;
   Cond heartbeat_cond;
   bool heartbeat_stop;
-  Mutex heartbeat_update_lock; // orders under heartbeat_lock
-  bool heartbeat_need_update;   ///< true if we need to refresh our heartbeat peers
+  std::atomic_bool heartbeat_need_update;
```
Comment (tchaikov) [inline src/osd/OSD.h]: i'd recommend use `std::atomic_flag` instead, `std::atomic<>` would use lock under the hood. but `std::atomic_flag` is lock free.

Comment (liupan1111) [inline src/osd/OSD.h]: cool....let me change it.

Comment (liupan1111) [inline src/osd/OSD.h]: @tchaikov, I tried std::atomic_flag, but I still prefer std::atomic<>. The reason is below:
1) 
#include <iostream>
#include <utility>
#include <atomic>

int main()
{
    std::atomic<bool> lock;
    std::cout << "free or not:" << std::endl;
    std::cout << lock.is_lock_free() << std::endl;
    return 0;
}

**Output:**
free or not:
1

So std::atomic<bool> is lock free.

2) there is not a function load for atomic_flag, so I have to do some change for heartbeat_peers_need_update return value and implement.

Do you agree?

**Comment by yuriw** [comment]
http://pulpito.ceph.com/yuriw-2016-12-13_18:35:58-rados-wip-yuri-testing_2016_12_14-distro-basic-smithi/

http://pulpito.ceph.com/yuriw-2016-12-14_01:52:35-rados-master---basic-smithi/

**Pull request merged**